libminifi/src/FlowController.cpp

```diff
@@ -150,6 +154,7 @@ bool FlowController::applyConfiguration(const std::string &source, const std::st
     this->root_ = std::move(prevRoot);
```

Comment:
won't unlocking at line 140 make it possible for this line to execute without a valid lock (due to an exception after the unlock)?

Reply:
I think in theory you are right, it could be unlocked if the exception happens after the unlock. Although the exception check is mostly meant for the loading of the new configuration, because after the unlock we only write some logs and set a few configuration values that shouldn't really throw any exceptions. Just in case I added the additional lock in 61844c0a17630875f7db41fcd329bedae7ef9df6

Reply:
After realizing that it could throw an exception if we lock multiple times I reverted this change. It shouldn't throw when it is unlocked so it should be okay.

Reply:
what was the problem that required manual locking/unlocking?

Reply:
As far as I remember it could cause a deadlock when the updating_ flag is checked from another thread and it's already set to false trying to acquire the lock afterwards, but it's still locked, so we have to unlock it before setting the update_ flag back to false.

Reply:
we should restructure this method to avoid manual locking, the whole lock protected region could get its own block, we should move the logging block out of the try block to the very end of this method

```
  updating_ = true;

  {
    std::lock_guard<std::recursive_mutex> flow_lock(mutex_);
    stop();
    unload();
    controller_map_->clear();
    clearResponseNodes();
    if (metrics_publisher_) {
      metrics_publisher_->clearMetricNodes();
    }
    auto prevRoot = std::move(this->root_);
    this->root_ = std::move(newRoot);
    processor_to_controller_.clear();
    initialized_ = false;
    bool started = false;
    try {
      load(std::move(root_), true);
      flow_update_ = true;
      started = start() == 0;
    } catch (...) {
      this->root_ = std::move(prevRoot);
      load(std::move(this->root_), true);
      flow_update_ = true;
    }
  }

  updating_ = false;

  if (started) {
    auto flowVersion = flow_configuration_->getFlowVersion();
    if (flowVersion) {
      logger_->log_debug("Setting flow id to %s", flowVersion->getFlowId());
      configuration_->set(Configure::nifi_c2_flow_id, flowVersion->getFlowId());
      configuration_->set(Configure::nifi_c2_flow_url, flowVersion->getFlowIdentifier()->getRegistryUrl());
    } else {
      logger_->log_debug("Invalid flow version, not setting");
    }
  }

  return started;
```

Reply:
It could happen when `executeOnAllComponents` or `executeOnComponent` was called from a metric while updating that's why the `updating_` check was added there.

Reply:
Good idea, updated in b53169315c1ec8d6222e8e55efab8631c75325c5